keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/vetos/RequiredSecurityGroupVeto.kt

```diff
@@ -0,0 +1,147 @@
+            }
+          }
+        }
+      }
+    }
+    return securityGroupMissingRegions
+  }
+
+  private suspend fun securityGroupExists(
+    account: String,
+    vpcName: String?,
+    region: String,
+    name: String
+  ): Boolean =
+    runCatching {
+      cloudDriver.getSecurityGroup(
+        DEFAULT_SERVICE_ACCOUNT,
+        account,
+        CLOUD_PROVIDER,
+        name,
+        region,
+        if (vpcName == null) null else cloudDriverCache.networkBy(vpcName, account, region).id
+      )
+    }
+      .onFailure { ex ->
+        if (ex is HttpException && ex.code() == 404) {
+          log.warn("security group $name not found in $account/$vpcName/$region")
+        } else {
+          log.error("error finding security group $name in $account/$vpcName/$region", ex)
+        }
+      }
+      .isSuccess
+
+  override fun messageFormat(): Map<String, Any> {
+    TODO("not implemented")
+  }
+
+  override fun passMessage(message: Map<String, Any>) {
+    TODO("not implemented")
+  }
+
+  override fun currentRejections(): List<String> {
+    TODO("not implemented")
+  }
+
+  override fun currentRejectionsByApp(application: String): List<String> {
+    TODO("not implemented")
+  }
+
+  data class SecurityGroupDependencies(
+    val securityGroupsInRegions: Map<String, Collection<String>>,
+    val account: String,
+    val subnet: String?
+  )
+
+  private fun ResourceSpec.toSecurityGroupDependencies() =
+    when (this) {
+      is LoadBalancerSpec ->
+        SecurityGroupDependencies(
+          securityGroupsInRegions = dependencies.securityGroupNames.associateWith {
+            locations.regions.map { it.name }
+          },
+          account = locations.account,
+          subnet = locations.subnet
+        )
+      is ClusterSpec -> {
+        val regions = locations.regions.map { it.name }
+        val defaultSecurityGroups = (defaults.dependencies?.securityGroupNames ?: emptySet())
+          .associateWith { regions }
+        val overrideSecurityGroups = overrides
+          .map { (region, spec) ->
+            spec.dependencies?.securityGroupNames?.associateWith { listOf(region) } ?: emptyMap()
+          }
+          .reduce { acc, map ->
+            val result = mutableMapOf<String, List<String>>()
+            (acc.keys + map.keys).forEach {
+              result[it] = (acc[it] ?: emptyList()) + (map[it] ?: emptyList())
+            }
+            result
+          }
+        val allSecurityGroups = listOf(defaultSecurityGroups, overrideSecurityGroups)
+          .reduce { acc, map ->
+            val result = mutableMapOf<String, List<String>>()
+            (acc.keys + map.keys).forEach {
+              result[it] = (acc[it] ?: emptyList()) + (map[it] ?: emptyList())
+            }
+            result
+          }
+        SecurityGroupDependencies(
+          securityGroupsInRegions = allSecurityGroups,
+          account = locations.account,
+          subnet = locations.subnet
+        )
+      }
```

Comment:
It sure feels like this should be shorter, but my brain is not firing on all cylinders today.